### PR TITLE
Wrapping k8s client with write filter and cache reader

### PR DESCRIPTION
**Pull request by hamersaw**
## Tracking issue
fixes https://github.com/flyteorg/flyte/issues/4730

## Why are the changes needed?
The previous attempt (https://github.com/flyteorg/flyte/pull/4733) mistakenly put he cache lookup into a loop until it was populated. This caused large CPU spikes.

## What changes were proposed in this pull request?
This PR bypasses the k8sClient internal caching logic and instead just wraps an uncached k8s client with a cache lookup for writes and a write through filter on `create` / `delete` operations.

## How was this patch tested?
local

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
_NA_

## Docs link
_NA_

### CODE REVIEW

File: cmd/single/start.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func startPropeller(ctx context.Context, cfg Propeller) error {
 			DefaultNamespaces: namespaceConfigs,
 		},
 		NewCache: executors.NewCache,
-		NewClient: executors.NewClient,
+		NewClient: executors.BuildNewClientFunc(50000, propellerScope),
```
Comment (honnix) [inline cmd/single/start.go:125]: I think if we don't see the `writeFilterSize` would be different, maybe for now we can just inline it. Wdyt?

File: flytepropeller/cmd/controller/cmd/webhook.go [hunk 2]
```diff
@@ -109,7 +109,7 @@ func runWebhook(origContext context.Context, propellerCfg *config.Config, cfg *w
 			DefaultNamespaces: namespaceConfigs,
 		},
 		NewCache:  executors.NewCache,
-		NewClient: executors.NewClient,
+		NewClient: executors.BuildNewClientFunc(50000, propellerScope),
```
Comment (honnix) [inline flytepropeller/cmd/controller/cmd/webhook.go:112]:
```suggestion
		NewClient: executors.BuildNewClientFunc(50000, webhookScope),
```

File: flytepropeller/pkg/controller/executors/kube.go [hunk 3]
```diff
@@ -33,107 +33,91 @@ var NewCache = func(config *rest.Config, options cache.Options) (cache.Cache, er
...
+		if options.Cache != nil && options.Cache.Reader != nil {
+			cacheReader = options.Cache.Reader
+			cachelessOptions.Cache = nil
 		}
 
-		options.Cache.Reader = reader
-	}
+		k8sClient, err := client.New(config, cachelessOptions)
+		if err != nil {
+			return k8sClient, err
```
Comment (honnix) [inline flytepropeller/pkg/controller/executors/kube.go:47]: It doesn't seem to be reasonable to return `k8sClient`, which could potentially be something built partially if `client.New` does something weird.

```suggestion
			return nil, err
```

File: cmd/single/start.go (hunk 1 above)
Comment (hamersaw) [reply cmd/single/start.go]: sounds good to me!

File: flytepropeller/cmd/controller/cmd/webhook.go (hunk 2 above)
Comment (hamersaw) [reply flytepropeller/cmd/controller/cmd/webhook.go]: good catch, thanks.

File: flytepropeller/pkg/controller/executors/kube.go (hunk 3 above)
Comment (hamersaw) [reply flytepropeller/pkg/controller/executors/kube.go]: done.

**Comment by honnix** [comment]
This looks good to me. I was struggling with unit test for this and I guess you might be the same. Not sure how to properly test it without some refactoring in place.

**Review by pingsutw — APPROVED** [review]

**Pull request merged**